compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp

```diff
@@ -60,6 +61,58 @@ struct LowerUnPackPattern : public OpRewritePattern<tensor::UnPackOp> {
   }
 };
 
+/// Folding trailing unit dims away from transpose op if they are not
+/// transposed. The decomposition stage is close to vectorization, this avoids
+/// trailing unit dims at vector types. It does not trigger issues at vector
+/// level because all the elements in a inner tiles are batch-handled. The
+/// strides of outer dims are still the same.
+struct FoldTrailingUnitTranspose
```

Comment:
Will this be needed with the n-D support that you posted upstream?

Reply:
This is needed, at least for now. There is a VectorToSCF stage before LowerVectorTranspose. The VectorToSCF patterns unroll all the vectors based on the last dim size. It destroys the form before lowering vector.transpose op.

I was looking for removing trailing unit dims for a while. After debugging for hours, I realized that it's infeasible for some cases. There are strides semantics after we move to memref's world, while vector ops are intended to read/write values in a continuous manner. One example is:

```
%v = transfer_read ... : memref<....>, vector<16x1xf32>
%t = vector.transpose %v : vector<16x1xf32> -> vector<1x16xf32>
vector.store %t ...
```

I end up with having the pattern which can navigate to the state we want better.

Reply:
I think I'm still missing something but I'm not totally convinced about adding and ah-hoc pattern to remove trailing dims from convolutions (tecnical_debt++). Not sure what has changed but patterns to lower n-D transposes to AVX2 was working at some point. What is different now?

> The VectorToSCF patterns unroll all the vectors based on the last dim size.

If this is a pass ordering problem, what is preventing us from moving VectorToSCF after the transpose lowering? Also, what happens for cases where the dimension is not unit (e.g., `1x16x16x3` and [0, 2, 1, 3]).

> There are strides semantics

Stride semantics would only impact the load/store part, not the vector computation itself as there are no strides in vectors. If we want to approach this properly, we should do something like:

```
%v = transfer_read ... : memref<....>, vector<16x1xf32>
%sh1 = vector.shape_cast %v : vector<16x1xf32> to vector<16xf32>
// Computation on vector<16xf32>
%sh2 = vector.shape_cast %v : vector<16xf32> to vector<16x1xf32>
vector.store %sh2 ...
```
 
```
%v = transfer_read ... : memref<....>, vector<16x1xf32> 
%t = vector.transpose %v : vector<16x1xf32> -> vector<1x16xf32> 
vector.store %t ...
```
We shouldn't have to worry about a transpose like this. We should actually turn it into a shape cast. I'm surprised we don't have a pattern for that already.

Reply:
There are many missing patterns to get the IR we want. I added the shape_cast things at vector level as you suggested offline. Here is what I get

Input:

```
      %subview = memref.subview %0[%2, %arg1] [16, 16] [1, 1] : memref<384x512xf32> to memref<16x16xf32, strided<[512, 1], offset: ?>>
      %expand_shape = memref.expand_shape %subview [[0, 1], [2, 3]] : memref<16x16xf32, strided<[512, 1], offset: ?>> into memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>>
      %3 = vector.transfer_read %expand_shape[%c0, %c0, %c0, %c0], %cst {in_bounds = [true, true, true, true]} : memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>>, vector<1x16x16x1xf32>
      %4 = vector.transpose %3, [0, 2, 1, 3] : vector<1x16x16x1xf32> to vector<1x16x16x1xf32>
      vector.transfer_write %4, %1[%arg0, %arg1, %c0, %c0] {in_bounds = [true, true, true, true]} : vector<1x16x16x1xf32>, memref<24x512x16x1xf32>
```

Adding the casting around the transpose op to fold all unit dims away:

```
        %subview = memref.subview %0[%2, %arg1] [16, 16] [1, 1] : memref<384x512xf32> to memref<16x16xf32, strided<[512, 1], offset: ?>>
        %expand_shape = memref.expand_shape %subview [[0, 1], [2, 3]] : memref<16x16xf32, strided<[512, 1], offset: ?>> into memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>>
        %subview_0 = memref.subview %expand_shape[0, 0, 0, 0] [1, 16, 16, 1] [1, 1, 1, 1] : memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>> to memref<1x16x16xf32, strided<[8192, 512, 1], offset: ?>>
        %subview_1 = memref.subview %subview_0[0, 0, 0] [1, 16, 16] [1, 1, 1] : memref<1x16x16xf32, strided<[8192, 512, 1], offset: ?>> to memref<16x16xf32, #map1>
        %3 = vector.transfer_read %subview_1[%c0, %c0], %cst {in_bounds = [true, true]} : memref<16x16xf32, #map1>, vector<16x16xf32>
        %4 = vector.shape_cast %3 : vector<16x16xf32> to vector<16x16x1xf32>
        %5 = vector.broadcast %4 : vector<16x16x1xf32> to vector<1x16x16x1xf32>
        %6 = vector.shape_cast %5 : vector<1x16x16x1xf32> to vector<16x16xf32>
        %7 = vector.transpose %6, [1, 0] : vector<16x16xf32> to vector<16x16xf32>
        %8 = vector.shape_cast %7 : vector<16x16xf32> to vector<1x16x16x1xf32>
        %9 = vector.extract %8[0] : vector<1x16x16x1xf32>
        vector.transfer_write %9, %1[%arg0, %arg1, %c0, %c0] {in_bounds = [true, true, true]} : vector<16x16x1xf32>, memref<24x512x16x1xf32>
```

With one line hack in [CastAwayTransferReadLeadingOneDim](https://github.com/llvm/llvm-project/blob/0a70d20c07a774a8b45dba0babd3515b36aa3590/mlir/lib/Dialect/Vector/Transforms/VectorDropLeadUnitDim.cpp#L228), I can remove some shape_cast and broadcast ops:

```
      scf.for %arg1 = %c0 to %c512 step %c16 {
        %subview = memref.subview %0[%2, %arg1] [16, 16] [1, 1] : memref<384x512xf32> to memref<16x16xf32, strided<[512, 1], offset: ?>>
        %expand_shape = memref.expand_shape %subview [[0, 1], [2, 3]] : memref<16x16xf32, strided<[512, 1], offset: ?>> into memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>>
        %subview_0 = memref.subview %expand_shape[0, 0, 0, 0] [1, 16, 16, 1] [1, 1, 1, 1] : memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>> to memref<1x16x16xf32, strided<[8192, 512, 1], offset: ?>>
        %subview_1 = memref.subview %subview_0[0, 0, 0] [1, 16, 16] [1, 1, 1] : memref<1x16x16xf32, strided<[8192, 512, 1], offset: ?>> to memref<16x16xf32, #map1>
        %3 = vector.transfer_read %subview_1[%c0, %c0], %cst {in_bounds = [true, true]} : memref<16x16xf32, #map1>, vector<16x16xf32>
        %4 = vector.transpose %3, [1, 0] : vector<16x16xf32> to vector<16x16xf32>
        %5 = vector.shape_cast %4 : vector<16x16xf32> to vector<1x16x16x1xf32>
        %6 = vector.extract %5[0] : vector<1x16x16x1xf32>
        vector.transfer_write %6, %1[%arg0, %arg1, %c0, %c0] {in_bounds = [true, true, true]} : vector<16x16x1xf32>, memref<24x512x16x1xf32>
```

I think what we want is not transfer_write a `16x16x1` vector and we want to get rid of casting, e.g., we want something like:

```
        %subview = memref.subview %0[%2, %arg1] [16, 16] [1, 1] : memref<384x512xf32> to memref<16x16xf32, strided<[512, 1], offset: ?>>
        %expand_shape = memref.expand_shape %subview [[0, 1], [2, 3]] : memref<16x16xf32, strided<[512, 1], offset: ?>> into memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>>
        %subview_0 = memref.subview %expand_shape[0, 0, 0, 0] [1, 16, 16, 1] [1, 1, 1, 1] : memref<1x16x16x1xf32, strided<[8192, 512, 1, 1], offset: ?>> to memref<1x16x16xf32, strided<[8192, 512, 1], offset: ?>>
        %subview_1 = memref.subview %subview_0[0, 0, 0] [1, 16, 16] [1, 1, 1] : memref<1x16x16xf32, strided<[8192, 512, 1], offset: ?>> to memref<16x16xf32, #map1>
        %3 = vector.transfer_read %subview_1[%c0, %c0], %cst {in_bounds = [true, true]} : memref<16x16xf32, #map1>, vector<16x16xf32>
        %4 = vector.transpose %3, [1, 0] : vector<16x16xf32> to vector<16x16xf32>
        %subview_2 = memref.subview %1 [0, 0, 0, 0] [24, 512, 16, 1] [1, 1, 1, 1] : memref<24x512x16x1xf32> to memref<24x512x16xf32>
        vector.transfer_write %4, %subview_2[%arg0, %arg1, %c0] {in_bounds = [true, true]} : vector<16x16xf32>, memref<24x512x16xf32>
```

There are two questions:

1. Is it able to apply my one line change upstream?
2. What's the right steps to fold the chain and make transfer_write drop all unit dims?

Reply:
I feel that this falls into the "file an issue for missing features, and land this as what it is" situation. If it only requires trivial changes or we are not using proper patterns, I can address all the concerns. But this is not the case. Something is off in the vector lowering black box. Fixing all the issues in the black box goes much beyond the scope of data tiling work. The workaround here is a "tech debt" for sure, but no other features will be built on top of it. We can simply delete the pattern if the vector lowering black box works better.

I feel that we should put "forming vector lowering into a better way" to our plan/roadmap, though I don't have concrete idea now. Otherwise we will likely fall into the same situation like this.